src/validation.cpp

```diff
@@ -2288,7 +2291,7 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
         while (disconnectpool->DynamicMemoryUsage() > MAX_DISCONNECTED_TX_POOL_SIZE * 1000) {
             // Drop the earliest entry, and remove its children from the mempool.
             auto it = disconnectpool->queuedTx.get<insertion_order>().begin();
-            m_mempool.removeRecursive(**it, MemPoolRemovalReason::REORG);
+            if (m_mempool) m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
```

Comment:
This `if (m_mempool)` condition can be at the level of the `if (disconnectpool)` above:

```diff
         return false;
     }
 
-    if (disconnectpool) {
+    if (m_mempool && disconnectpool) {
         // Save transactions to re-add to mempool at end of reorg
         for (auto it = block.vtx.rbegin(); it != block.vtx.rend(); ++it) {
             disconnectpool->addTransaction(*it);
@@ -2291,7 +2291,7 @@ bool CChainState::DisconnectTip(BlockValidationState& state, DisconnectedBlockTr
         while (disconnectpool->DynamicMemoryUsage() > MAX_DISCONNECTED_TX_POOL_SIZE * 1000) {
             // Drop the earliest entry, and remove its children from the mempool.
             auto it = disconnectpool->queuedTx.get<insertion_order>().begin();
-            if (m_mempool) m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
+            m_mempool->removeRecursive(**it, MemPoolRemovalReason::REORG);
             disconnectpool->removeEntry(it);
         }
     }
```

That would potentially save some work updating the `disconnectpool` data structure, which will never get used.

Even better (but not necessarily part of this PR) would be to only create a `DisconnectedBlockTransactions` if `m_mempool` is non-null, and then update ` UpdateMempoolForReorg()` and ` DisconnectedBlockTransactions()` to take optional `DisconnectedBlockTransactions*`. There's no need to create and maintain a `DisconnectedBlockTransactions` if we're never going to use it.

Reply:
Yup, also a good call. Fixed. Initially wasn't sure if there was some other purpose we might be using the disconnectpool for (since it's sometimes passed in as a mutable parameter) but after looking around it's clearly just used for mempool maintenance.